compiler/rustc_mir_build/src/build/expr/as_rvalue.rs

```diff
@@ -197,16 +198,42 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 // create all the steps directly in MIR with operations all backends need to support anyway.
                 let (source, ty) = if let ty::Adt(adt_def, ..) = source.ty.kind() && adt_def.is_enum() {
                     let discr_ty = adt_def.repr().discr_type().to_ty(this.tcx);
-                    let temp = unpack!(block = this.as_temp(block, scope, source, Mutability::Not));
+                    let layout = this.tcx.layout_of(this.param_env.and(source.ty));
                     let discr = this.temp(discr_ty, source.span);
+                    let temp = unpack!(block = this.as_temp(block, scope, source, Mutability::Not));
                     this.cfg.push_assign(
                         block,
                         source_info,
                         discr,
                         Rvalue::Discriminant(temp.into()),
                     );
+                    let (op,ty) = (Operand::Move(discr), discr_ty);
+
+                    if let Abi::Scalar(scalar) = layout.unwrap().abi{
+                        if let Primitive::Int(..) = scalar.primitive(){
+                           let end_range = scalar.valid_range(&this.tcx).end;
```

Comment:
If the range start is larger than the end, that means we have a wraparound range. In that case the assume will be wrong. Either check for that case or generalize the logic to handle range start, too, as well as wraparound ranges.

Should probably add a test for it, too